src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs

```diff
@@ -1090,6 +1090,70 @@ public async Task EchoServerStreaming_DifferentMessageSize_Success(int messageSi
             connection.Dispose();
         }
 
+        [Fact]
+        [OuterLoop("Uses Task.Delay")]
+        public async Task RequestContentStreaming_Timeout_BothClientAndServerReceiveCancellation()
+        {
+            if (UseQuicImplementationProvider == QuicImplementationProviders.Mock)
+            {
+                return;
+            }
+
+            var message = new byte[1024];
+            var random = new Random(0);
+            random.NextBytes(message);
+
+            using Http3LoopbackServer server = CreateHttp3LoopbackServer();
+            Http3LoopbackConnection connection = null;
+            Http3LoopbackStream serverStream = null;
+
+            Task serverTask = Task.Run(async () =>
+            {
+                connection = (Http3LoopbackConnection)await server.EstablishGenericConnectionAsync();
+                serverStream = await connection.AcceptRequestStreamAsync();
+
+                await serverStream.HandleRequestAsync().WaitAsync(TimeSpan.FromSeconds(60));
+            });
+
+            StreamingHttpContent requestContent = new StreamingHttpContent();
+
+            using HttpClient client = CreateHttpClient();
+            client.Timeout = TimeSpan.FromSeconds(0.3); // set small timeout
+            using HttpRequestMessage request = new()
+            {
+                Method = HttpMethod.Post,
+                RequestUri = server.Address,
+                Version = HttpVersion30,
+                VersionPolicy = HttpVersionPolicy.RequestVersionExact,
+                Content = requestContent
+            };
+
+            var responseTask = client.SendAsync(request).WaitAsync(TimeSpan.FromSeconds(60));
+
+            Stream requestStream = await requestContent.GetStreamAsync();
+            // Send headers
+            await requestStream.FlushAsync();
+
+            await requestStream.WriteAsync(message).AsTask().WaitAsync(TimeSpan.FromSeconds(10));
+            await requestStream.FlushAsync();
+            
+            await Task.Delay(TimeSpan.FromSeconds(2)); // longer than client.Timeout
+
+            // client is properly canceled on timeout
+            await Assert.ThrowsAsync<ObjectDisposedException>(() => requestStream.WriteAsync(message).AsTask().WaitAsync(TimeSpan.FromSeconds(10)));
```

Comment:
Why ODE here? Is this some of our filed, punted, unresolved issues?

Reply:
That's what happens to HTTP/2 as well. The code for sending the content is like
```
using (var httpWriteStream = new Http(2|3)WriteStream(httpRequestStream))
{
    await content.CopyToAsync(httpWriteStream);
}
```
So as soon as an error happens in CopyToAsync (or it successfully finishes), httpWriteStream is disposed.
The way how CopyToAsync is implemented and how it treats exceptions inside is up to specific Content implementation. The one I used just provided the httpWriteStream outside to the user, so after it gets disposed between two calls, the ODE happens.

Reply:
Correct me if I'm wrong please.
So user normally doesn't see ODE. We observe it here because we're directly touching `Http3WriteStream` outside of the normal content sending flow. And in case of duplex, the `cancellationToken` doesn't apply outside of `SendAsync` so the write stream won't get disposed under user's hands just by a timeout/cancellation.

Reply:
That's right.